crates/sui-display/src/v2/lexer.rs

```diff
@@ -0,0 +1,846 @@
+
+        use Token as T;
+        Some(match bytes.first()? {
+            b'@' => self.take(T::At, 1),
+
+            b':' if bytes.get(1) == Some(&b':') => self.take(T::CColon, 2),
+
+            b':' => self.take(T::Colon, 1),
+
+            b',' => self.take(T::Comma, 1),
+
+            b'.' => self.take(T::Dot, 1),
+
+            b'0' if bytes.get(1) == Some(&b'x')
+                && bytes.get(2).is_some_and(|b| is_valid_hex_byte(*b)) =>
+            {
+                self.advance(2);
+                self.take_until(T::NumHex, |c| !is_valid_hex_byte(c))
+            }
+
+            b'0'..=b'9' => self.take_until(T::NumDec, |c| !is_valid_decimal_byte(c)),
+
+            b'a'..=b'z' | b'A'..=b'Z' => {
+                self.take_until(T::Ident, |c| !is_valid_identifier_byte(c))
+            }
+
+            b'<' => self.take(T::LAngle, 1),
+
+            b'{' => self.take(T::LBrace, 1),
+
+            b'[' => self.take(T::LBracket, 1),
+
+            b'(' => self.take(T::LParen, 1),
+
+            b'|' => self.take(T::Pipe, 1),
+
+            b'#' => self.take(T::Pound, 1),
+
+            b'>' => self.take(T::RAngle, 1),
+
+            b'}' => {
+                self.mode = Mode::Text;
+                self.take(T::RBrace, 1)
+            }
+
+            b']' => self.take(T::RBracket, 1),
+
+            b')' => self.take(T::RParen, 1),
+
+            b'\'' => {
+                // Set the escaped indicator to true initially so we don't interpret the starting
+                // quote as an ending quote.
+                let mut escaped = true;
+                for (i, b) in self.src.bytes().enumerate() {
+                    if escaped {
+                        escaped = false;
+                    } else if b == b'\\' {
+                        escaped = true;
+                    } else if b == b'\'' {
+                        self.advance(1);
+                        let content = self.take(T::String, i - 1);
+                        self.advance(1);
+                        return Some(content);
+                    }
+                }
+
+                // Reached the end of the byte stream and didn't find a closing quote -- treat the
+                // partial string as an unexpected token.
+                self.take(T::Unexpected, self.src.len())
+            }
+
+            // Explicitly tokenize whitespace.
+            _ if self.src.chars().next().is_some_and(char::is_whitespace) => self.take(
+                Token::Whitespace,
+                self.src
+                    .find(|c: char| !c.is_whitespace())
+                    .unwrap_or(self.src.len()),
+            ),
+
+            // If the next byte cannot be recognized, extract the next (potentially variable
+            // length) character, and indicate that it is an unexpected token.
+            _ => {
+                let mut indices = self.src.char_indices();
+                indices.next(); // skip the first character
```

Comment:
As discussed we need to handle the case where `self.src` may not contain any valid utf8 chars in at this point. Possibly `is_char_boundary` to check for this before creating the iterator.